cmd/e2e-test/carbon_clickhouse.go

```diff
@@ -0,0 +1,124 @@
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path"
+	"path/filepath"
+	"strconv"
+	"syscall"
+	"text/template"
+
+	"github.com/phayes/freeport"
+)
+
+type CarbonClickhouse struct {
+	Binary    string `toml:"binary"`
+	ConfigTpl string `toml:"template"`
+
+	storeDir   string    `toml:"-"`
+	configFile string    `toml:"-"`
+	address    string    `toml:"-"`
+	cmd        *exec.Cmd `toml:"-"`
+}
+
+func (c *CarbonClickhouse) Start(clickhouseAddr string) error {
+	if c.cmd != nil {
+		return fmt.Errorf("carbon-clickhouse already started")
+	}
+
+	if len(c.Binary) == 0 {
+		c.Binary = "./carbon-clickhouse"
+	}
+	if len(c.ConfigTpl) == 0 {
+		return fmt.Errorf("carbon-clickhouse config template not set")
+	}
+
+	var err error
+
+	c.storeDir, err = ioutil.TempDir("", "carbon-clickhouse")
+	if err != nil {
+		return err
+	}
+	port, err := freeport.GetFreePort()
+	if err != nil {
+		c.Cleanup()
+		return err
+	}
+	c.address = "127.0.0.1:" + strconv.Itoa(port)
+
+	name := filepath.Base(c.ConfigTpl)
+	tmpl, err := template.New(name).ParseFiles(c.ConfigTpl)
+	if err != nil {
+		c.Cleanup()
+		return err
+	}
+	param := struct {
+		CLICKHOUSE_ADDR string
+		CCH_ADDR        string
+		CCH_STORE_DIR   string
+	}{
+		CLICKHOUSE_ADDR: clickhouseAddr,
+		CCH_ADDR:        c.address,
+		CCH_STORE_DIR:   c.storeDir,
+	}
```

Comment:
```suggestion
	}{clickhouseAddr, c.address, c.storeDir}
```

Does it, maybe, look a little bit better?

Reply:
It's a anonymous struct, without struct definition in-place (or declare named struct) it's not work.

Reply:
Thanks, will know it :+1: